lightning/src/ln/onion_utils.rs

```diff
@@ -442,30 +429,58 @@ where L::Target: Logger {
 		let mut hmac = HmacEngine::<Sha256>::new(&um);
 		hmac.input(&err_packet.encode()[32..]);
 
+		// The failing hop includes either the inbound channel to the recipient or the outbound channel
+		// from the current hop (i.e., the next hop's inbound channel).
+		let num_blinded_hops = path.blinded_tail.as_ref().map_or(0, |t| t.hops.len());
+		is_from_final_node = route_hop_idx + 1 == path.hops.len() &&
+			(path.blinded_tail.is_none() || num_blinded_hops == 1);
+
 		if !fixed_time_eq(&Hmac::from_engine(hmac).into_inner(), &err_packet.hmac) { return }
 		let error_code_slice = match err_packet.failuremsg.get(0..2) {
 			Some(s) => s,
-			None => {
+			None if route_hop_opt.is_some() => {
 				// Useless packet that we can't use but it passed HMAC, so it definitely came from the peer
 				// in question
 				let network_update = Some(NetworkUpdate::NodeFailure {
-					node_id: route_hop.pubkey,
+					node_id: route_hop_opt.as_ref().unwrap().pubkey,
 					is_permanent: true,
 				});
-				let short_channel_id = Some(route_hop.short_channel_id);
+				let short_channel_id = Some(route_hop_opt.as_ref().unwrap().short_channel_id);
 				res = Some((network_update, short_channel_id, !is_from_final_node));
 				return
+			},
+			None => return, // Useless packet from a blinded path hop
```

Comment:
I don't quite understand this change. Doesn't skipping the check on `route_hop_opt` earlier result in executing a bunch of cryptographic functions above just to `return` here whereas before they would have been skipped?

Reply:
We can't return early anymore because even in the blinded case, we need to set `error_code`/`error_data` and `is_from_final_node` (which affects `PaymentPathFailed::payment_failed_permanently`). So now in the blinded case we return *after* those fields are set, if that makes sense.

Reply:
Could you expand some of these commit messages describe the _why_ of the changes? Would help with understanding things like this more.